src/libstore/ssh.cc

```diff
@@ -62,21 +62,58 @@ std::unique_ptr<SSHMaster::Connection> SSHMaster::startCommand(const std::string
         if (logFD != -1 && dup2(logFD, STDERR_FILENO) == -1)
             throw SysError("duping over stderr");
 
-        Strings args;
+        Strings execargs;
 
         if (fakeSSH) {
-            args = { "bash", "-c" };
+            execargs.push_back(command);
+            execargs.splice(execargs.end(), args);
         } else {
-            args = { "ssh", host.c_str(), "-x" };
-            addCommonSSHOpts(args);
+            std::string escapedCommand = shellEscape(command);
+
+            std::string script = {escapedCommand};
+
+            for (std::string arg: args) {
+                script.push_back(' ');
+                script.append(shellEscape(arg));
+            }
+
+            // This is really ugly, but it's less ugly then trying to catch the
+            // error thrown by this process when ssh it can't find the command.
+            // We would have to first of all catch this error (which is not
+            // trivial, since it will only be thrown to the parent process
+            // during blocking communication), then redirect the stderr, search
+            // it for "command not found" and then ssh again and try to find the
+            // command.
+            script.append(
+                fmt(
+                    " || ( \
+                    code=$?;\
+                    if ! [ $code -eq 127 ]; then exit $code; fi;\
+                    PATH=\"$PATH:$HOME/.nix-profile/bin:/nix/var/nix/profiles/default/bin\";\
+                    cmd=\"$(command -v %s)\";\
+                    if [ -n \"$cmd\" ]; then\
+                        printf \"Executable %s not found in PATH on the remote host, but it was found in %%s.\\nPerhaps try adding '\\e[35;1m&remote-program=%%s\\e[0m' at the end of your SSH url.\n\" \"$cmd\" \"$cmd\" > /dev/stderr; \
+                    else\
+                        printf \"Executable %s was not found on the remote host. Are you sure you have Nix installed there?\n\" > /dev/stderr;\
+                    fi;\
+                    exit $code;\
+                    )",
+                    escapedCommand,
+                    escapedCommand,
+                    escapedCommand
+                )
+            );
```

Comment:
I think it would be much cleaner (though a bit slower) to have that logic on the client side

Reply:
Which client side, though? As far as I understand, at this stage we haven't found the "client" yet.

Reply:
Ah, never mind, confusing client and server here. Do you want me to do what I described in the comment? Because I think it would be a lot more hacky due to having to catch an exception from another process :(

Reply:
Duh, I didn't read the comment :man_facepalming: 

Yeah, that would be hacky indeed, but afaik `ssh` forwards the return code, so checking for a return code `127` should work well-enough I think. But maybe that would still make it somewhat ugly :man_shrugging:

Reply:
Hm, that sounds good. Where would we actually `catch` the error though? This is a separate process from the parent one. There are a couple of options:

1. Wait some time to let ssh connect and fail if there's no such command, only then continue with the communcation. This is _not great_, obviously.
2. Write (in a blocking way) _something_ to the pipe, `catch` there. It's not clear what _something_ should be, it sounds hacky, and I'm not sure the exception thrown (probably "unexpected EOF") will actually have the exit code of ssh.

Reply:
Mh, I see. I'll defer this to @Ericson2314 as the owner of that PR :stuck_out_tongue:

Reply:
Yeah I would advocate for a hybrid approach where we do the `command -v %s` thing, and print out sentinel values `NIX YES` or `NIX NO` or whatever, but the rest of the logic is in C++. Does that make sense?

Reply:
You mean do the `command -v` dance before we run the command, then capture the output? Otherwise I can't see how it would work, given the limitations outlined in the previous comment.

Reply:
I mean something like this:

1. ssh normally
2. if failled, warn something like "${remoteProgram} on remote end not found, troubleshooting..."
3. begin trouble shooting loop over various additional path entries to try
   1. ssh again to run `PATH=$PATH:foo command -v; echo "COMMAND RESULT $!"` or similar, for some `foo`.
   2. parse the result in C++
   3. If something was found, print instructions to use for how to fix, and break out of loop.
 4. else If nothing was found, print something like "troubleshooting failed" and the original error.

Reply:
> 2. if failled, warn something like "${remoteProgram} on remote end not found, troubleshooting..."

It's not clear to me how we would catch this, and then decide that it failed because of a missing program and not some other problem.

> 3. if failled, warn something like "${remoteProgram} on remote end not found, troubleshooting..."

TBH this sounds like it would be even more hacky.

I don't think I want to spend the time to implement this, since I feel like it would be an ugly hackfest regardless.

Reply:
> It's not clear to me how we would catch this, and then decide that it failed because of a missing program and not some other problem.

That is why we are doing `echo "COMMAND RESULT $!"`, so we can grep for a specific sentinel value and not get confused about other errors.

> TBH this sounds like it would be even more hacky.

There is no fancy condition for that warning. that is simply to let the user know we are going to start poking at the remote some more. If we *do* hit another error and don't get `COMMAND RESULT <number>` back, it allows us to say something like "troubleshooting failed" indicating the issue is something else than what we are aware of.

Reply:
The point of all this is to keep the *happy* path very clean and simple, no messing with the `PATH` or anything else, but the *sad* path is free to try random stuff in a non-normative way to assist the user with debugging the issue.

Reply:
> The point of all this is to keep the happy path very clean and simple, no messing with the PATH or anything else, but the sad path is free to try random stuff in a non-normative way to assist the user with debugging the issue.


I think the happy path (before catching the error with `||`) is exactly as before with my solution as well. This ugly mess only happens when running the usual command fails.

Reply:
I don't mind there being a second ssh. The user can always control-C out if they don't want to do it; the "we're going to try trouble shooting now" message lets them know what's going on.